File: mobile/src/api/eventLocalReminders.ts
import type { Reminder } from '@aperio/shared';

import CalFfi from '../../modules/cal-ffi/src/CalFfiModule';
import { scheduleBackgroundPush } from './syncTriggers';

/**
 * Reminders Aperio keeps for ONE event and tells no provider about
 * (migration 0043).
 *
 * A reminder normally rides on the event: Aperio writes it into the
 * appointment, the provider stores it, and every other client of that calendar
 * rings too — the iOS Calendar app, a voice assistant reading the account out
 * loud. These do not. They fire in Aperio, travel to the user's other devices
 * through Aperio's own sync, and reach nobody else on a shared calendar.
 *
 * Rows are keyed by the SERIES master id, like colour overrides and group
 * membership: a recurring appointment is reminded of as a series.
 */
export interface EventLocalReminders {
  calendar_id: string;
  event_id: string;
  /** An empty list is a decision ("none of my own here"), not an absence. */
  reminders: Reminder[];
  /** The title the event had when this was set. Half of the SIGNATURE, so the
   *  row can find its event again after the provider remints the id. */
  title: string;
  /** The start it had then. The other half. */
  starts_at: string;
  updated_at: string;
}

/** Every row. Small by nature — one per event with a private reminder. */
export const listEventLocalReminders = async (): Promise<EventLocalReminders[]> =>
  JSON.parse(await CalFfi.eventLocalRemindersJson()) as EventLocalReminders[];

/** Write one event's private reminders. `title`/`startsAt` are the event's
 *  CURRENT signature. An empty list is stored, not deleted. */
export const setEventLocalReminders = async (payload: {
  calendar_id: string;
  event_id: string;
  reminders: Reminder[];
  title: string;
  starts_at: string;
}): Promise<EventLocalReminders> => {
  const row = JSON.parse(
    await CalFfi.setEventLocalRemindersJson(
      payload.calendar_id,
      payload.event_id,
      JSON.stringify(payload.reminders),
      payload.title,
      payload.starts_at,
    ),
  ) as EventLocalReminders;
  // A synced record: push it now rather than at the next periodic round, the
  // way every other mobile mutation does.
  scheduleBackgroundPush();
  return row;
};

/** Point a row at the id its event carries now.
 *
 *  A repair of Aperio's own bookkeeping — the same appointment before and
 *  after — so it is applied silently by whichever view noticed and never
 *  announced as a change the user made. */
export const healEventLocalReminders = (payload: {
  calendar_id: string;
  old_event_id: string;
  new_event_id: string;
}): Promise<boolean> =>
  CalFfi.healEventLocalReminders(
    payload.calendar_id,
    payload.old_event_id,
    payload.new_event_id,
  );

/** Write down what the event looks like now, so the signature keeps matching
 *  after the user renames or moves the appointment. Silent, like the repair. */
export const refreshEventLocalReminderSignature = (payload: {
  calendar_id: string;
  event_id: string;
  title: string;
  starts_at: string;
}): Promise<void> =>
  CalFfi.refreshEventLocalReminderSignature(
    payload.calendar_id,
    payload.event_id,
    payload.title,
    payload.starts_at,
  );

File: crates/cal-core/src/event_local_reminders.rs
//! Reminders Aperio keeps for one event and tells no provider about.
//!
//! A reminder normally rides ON the event, so every client of the calendar
//! rings. These do not: they live in Aperio, travel between the user's own
//! devices through Aperio's sync, and reach nobody else on a shared calendar.
//! See migration `0043_event_local_reminders.sql`.
//!
//! The type lives here rather than in the host because it crosses crates, the
//! same arrangement [`crate::EventGroup`] has: the host writes it, the sync
//! applier reads it off the wire, and the local adapter stores what arrives.

use serde::{Deserialize, Serialize};

use crate::Reminder;

/// One event's Aperio-only reminders.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct EventLocalReminders {
    pub calendar_id: String,
    /// Series master id — a recurring appointment is reminded of as a series,
    /// exactly as its own reminders are.
    pub event_id: String,
    /// An empty list is a decision ("none of my own here"), not an absence:
    /// the row stays so a peer that has not heard yet has something to lose
    /// against. See the migration.
    pub reminders: Vec<Reminder>,
    /// The title the event had when this was set. Half of the SIGNATURE, and
    /// never shown: what the user reads comes from the event itself, which may
    /// since have been renamed.
    pub title: String,
    /// The start it had then. The other half.
    pub starts_at: String,
    pub updated_at: String,
}

impl EventLocalReminders {
    /// Whether this row asks for anything at all.
    pub fn is_empty(&self) -> bool {
        self.reminders.is_empty()
    }
}
